### PR TITLE
Fix compile panic on non existent type return

### PR DESCRIPTION
**Pull request by JoshBrudnak**
Reverted the change https://github.com/rust-lang/rust/commit/28a76a90009d605349babcd2755962ab93913327#diff-4ed25c00aceb84666fca639cf8101c7cL1069 which was panicking when returning a type that cannot be found in the current scope and added testing for the compile error.

For example:
```rust
fn addition() -> Wrapper<impl A> {}
```
Where Wrapper is undefined in the scope.

### CODE REVIEW

File: src/librustc_typeck/collect.rs [hunk 1]
```diff
@@ -25,41 +25,43 @@
 //! crate as a kind of pass. This should eventually be factored away.
 
 use astconv::{AstConv, Bounds};
-use lint;
```
Comment (oli-obk) [inline src/librustc_typeck/collect.rs]: Please separate the rustfmt changes into their own commit

File: src/test/compile-fail/issue-53300.rs [hunk 2]
```diff
@@ -0,0 +1,22 @@
+// Copyright 2012-2014 The Rust Project Developers. See the COPYRIGHT
```
Comment (estebank) [inline src/test/compile-fail/issue-53300.rs]: Can you turn this into an `ui` test instead? We're slowly moving away from `compile-fail`.

Comment (JoshBrudnak) [reply src/test/compile-fail/issue-53300.rs]: K, will do. Thanks. :+1:

**Comment by estebank** [comment]
@JoshBrudnak please don't use merge, use rebase for clean patches instead. Also, the merge failed due to #52953:

```
[00:13:40] error[E0432]: unresolved import `syntax::codemap`
[00:13:40]   --> librustc_typeck/collect.rs:46:13
[00:13:40]    |
[00:13:40] 46 | use syntax::codemap::Spanned;
[00:13:40]    |             ^^^^^^^ Could not find `codemap` in `syntax`
[00:13:40] 
[00:13:46] error: aborting due to previous error
```

**Comment by JoshBrudnak** [comment]
Rebased to fix the merge conflict.

**Pull request merged**